### PR TITLE
Pt 162635989 internal data representation

### PR DESCRIPTION
**Pull request by happi**
Initial version of Fate data representation and serialization and deserialization.

### CODE REVIEW

File: apps/aefate/src/aefa_data.hrl [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+-define(FATE_STRING_SIZE(X), (byte_size(X))).
+-define(FATE_TRUE,  true).
+-define(FATE_FALSE, false).
+-define(FATE_NIL,   []).
+-define(FATE_VOID,  void).
+-define(FATE_EMPTY_STRING, <<>>).
+-define(FATE_STRING(S), S).
+-define(FATE_VARIANT(Tag,T), {variant, Tag, T}).
+
+-define(MAKE_FATE_INTEGER(X), X).
```
Comment (gorillainduction) [inline apps/aefate/src/aefa_data.hrl]:
```suggestion
-define(MAKE_FATE_INTEGER(X), begin [error(illegal_integer, X) || not ?IS_FATE_INTEGER(X)], X end).
```

File: apps/aefate/src/aefa_data.hrl [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+-define(MAKE_FATE_LIST(X), X).
```
Comment (gorillainduction) [inline apps/aefate/src/aefa_data.hrl]:
```suggestion
-define(MAKE_FATE_LIST(X), begin [error(illegal_list, X} || not ?IS_FATE_LIST(X)], X end).
```

File: apps/aefate/src/aefa_data.hrl [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+-define(MAKE_FATE_MAP(X),  X).
```
Comment (gorillainduction) [inline apps/aefate/src/aefa_data.hrl]:
```suggestion
 -define(MAKE_FATE_MAP(X),  begin [error(illegal_map, X} || not ?IS_FATE_MAP(X)], X end).
```

File: apps/aefate/src/aefa_data.hrl [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+-define(MAKE_FATE_STRING(X),  X).
```
Comment (gorillainduction) [inline apps/aefate/src/aefa_data.hrl]:
```suggestion
-define(MAKE_FATE_STRING(X),  begin [error(illegal_string, X} || not ?IS_FATE_STRING(X)], X end).
```

File: apps/aefate/src/aefate.app.src [hunk 5]
```diff
@@ -0,0 +1,16 @@
+{application, aefate,
+ [{description, "Fate: Fast aeternity transaction engine"},
+  {vsn, "0.0.1"},
+  {registered, []},
+  {mod, { aefate_app, []}},
```
Comment (gorillainduction) [inline apps/aefate/src/aefate.app.src]: This application is a library application and should not have an application callback module.

File: apps/aefate/src/aefate_app.erl [hunk 6]
```diff
@@ -0,0 +1,29 @@
+%%%-------------------------------------------------------------------
```
Comment (gorillainduction) [inline apps/aefate/src/aefate_app.erl]: This module is unnecessary in a library application.

File: apps/aefate/src/aefa_data.hrl (hunk 1 above)
Comment (happi) [reply apps/aefate/src/aefa_data.hrl]: Hmm but an int is not a FATE_INTEGER before it is turned into one.
Possibly .. || not is_integer(X)...

**Pull request merged**